File: internal/services/bigquery/sql_eval.go
package bigquery

import (
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

func executeParsedQuery(schema tableSchema, rows []storedRow, query simpleSelectQuery) (queryExecutionResult, error) {
	selectedFields, err := fieldsForQuery(schema, query.SelectedFields)
	if err != nil {
		return queryExecutionResult{}, err
	}
	filtered := make([]storedRow, 0, len(rows))
	for _, row := range rows {
		matches, err := rowMatchesQuery(row, query)
		if err != nil {
			return queryExecutionResult{}, err
		}
		if matches {
			filtered = append(filtered, row)
		}
	}
	if query.Aggregate.Function != "" {
		if query.GroupBy != "" {
			return executeGroupedAggregateQuery(filtered, schema, query)
		}
		return executeAggregateQuery(filtered, schema, query)
	}
	if query.OrderBy != "" {
		sort.SliceStable(filtered, func(i, j int) bool {
			left := filtered[i].JSON[query.OrderBy]
			right := filtered[j].JSON[query.OrderBy]
			cmp := compareRawValues(left, right)
			if query.OrderDesc {
				return cmp > 0
			}
			return cmp < 0
		})
	}
	if query.Offset > 0 {
		if query.Offset >= len(filtered) {
			filtered = nil
		} else {
			filtered = filtered[query.Offset:]
		}
	}
	if query.Limit >= 0 && query.Limit < len(filtered) {
		filtered = filtered[:query.Limit]
	}
	responseRows := make([]tableDataRow, 0, len(filtered))
	for _, row := range filtered {
		responseRows = append(responseRows, tableDataRow{F: formatRowValues(row.JSON, selectedFields)})
	}
	return queryExecutionResult{Fields: selectedFields, Rows: responseRows}, nil
}

func queryResultRowsToStoredRows(result queryExecutionResult) []storedRow {
	rows := make([]storedRow, 0, len(result.Rows))
	for _, row := range result.Rows {
		values := make(map[string]json.RawMessage, len(result.Fields))
		for i, field := range result.Fields {
			var value any
			if i < len(row.F) {
				value = row.F[i].V
			}
			raw, err := json.Marshal(value)
			if err != nil {
				raw = []byte("null")
			}
			values[field.Name] = json.RawMessage(raw)
		}
		rows = append(rows, storedRow{JSON: values})
	}
	return rows
}

func executeAggregateQuery(rows []storedRow, schema tableSchema, query simpleSelectQuery) (queryExecutionResult, error) {
	fieldName := query.Aggregate.Alias
	if fieldName == "" {
		fieldName = "f0_"
	}

	field, hasField := aggregateField(schema, query.Aggregate)

	switch query.Aggregate.Field {
	case "*":
		if query.Aggregate.Function != "COUNT" {
			return queryExecutionResult{}, fmt.Errorf("%s requires a field", query.Aggregate.Function)
		}
		data, _ := json.Marshal(strconv.Itoa(len(rows)))
		return queryExecutionResult{
			Fields: []tableFieldSchema{{
				Name: fieldName,
				Type: "INTEGER",
				Mode: "NULLABLE",
			}},
			Rows: []tableDataRow{{
				F: []tableCell{{V: rawValueForResponse(data)}},
			}},
		}, nil
	default:
		if !hasField {
			return queryExecutionResult{}, fmt.Errorf("aggregate field %q does not exist", query.Aggregate.Field)
		}
	}

	switch query.Aggregate.Function {
	case "COUNT":
		count := 0
		for _, row := range rows {
			raw, ok := row.JSON[query.Aggregate.Field]
			if ok && !isJSONNull(raw) {
				count++
			}
		}
		return singleAggregateResult(fieldName, "INTEGER", strconv.Itoa(count)), nil
	case "SUM":
		if !isNumericField(field) {
			return queryExecutionResult{}, fmt.Errorf("SUM requires a numeric field")
		}
		sum, count, err := sumAggregate(rows, query.Aggregate.Field, isIntegerField(field))
		if err != nil {
			return queryExecutionResult{}, err
		}
		if count == 0 {
			return singleAggregateNullResult(fieldName, aggregateNumericType(field)), nil
		}
		return singleAggregateResult(fieldName, aggregateNumericType(field), sum), nil
	case "AVG":
		if !isNumericField(field) {
			return queryExecutionResult{}, fmt.Errorf("AVG requires a numeric field")
		}
		sum, count, err := floatAggregate(rows, query.Aggregate.Field)
		if err != nil {
			return queryExecutionResult{}, err
		}
		if count == 0 {
			return singleAggregateNullResult(fieldName, "FLOAT"), nil
		}
		return singleAggregateResult(fieldName, "FLOAT", strconv.FormatFloat(sum/float64(count), 'f', -1, 64)), nil
	case "MIN", "MAX":
		raw, ok := minMaxAggregate(rows, query.Aggregate.Field, query.Aggregate.Function)
		if !ok {
			return singleAggregateNullResult(fieldName, defaultString(field.Type, "STRING")), nil
		}
		return queryExecutionResult{
			Fields: []tableFieldSchema{{
				Name: fieldName,
				Type: defaultString(field.Type, "STRING"),
				Mode: "NULLABLE",
			}},
			Rows: []tableDataRow{{
				F: []tableCell{{V: rawValueForResponse(raw)}},
			}},
		}, nil
	default:
		return queryExecutionResult{}, fmt.Errorf("unsupported aggregate function")
	}
}

func executeGroupedAggregateQuery(rows []storedRow, schema tableSchema, query simpleSelectQuery) (queryExecutionResult, error) {
	fields, err := groupedAggregateDryRunFields(schema, query)
	if err != nil {
		return queryExecutionResult{}, err
	}
	groups := make(map[string][]storedRow)
	groupValues := make(map[string]json.RawMessage)
	for _, row := range rows {
		raw, ok := row.JSON[query.GroupBy]
		if !ok {
			raw = json.RawMessage("null")
		}
		key := string(raw)
		groups[key] = append(groups[key], row)
		groupValues[key] = raw
	}
	keys := make([]string, 0, len(groups))
	for key := range groups {
		keys = append(keys, key)
	}
	sort.Slice(keys, func(i, j int) bool {
		cmp := compareRawValues(groupValues[keys[i]], groupValues[keys[j]])
		if query.OrderDesc {
			return cmp > 0
		}
		return cmp < 0
	})
	if query.OrderBy != "" && query.OrderBy != query.GroupBy {
		return queryExecutionResult{}, fmt.Errorf("ORDER BY supports grouped field only for GROUP BY queries")
	}
	if query.Offset > 0 {
		if query.Offset >= len(keys) {
			keys = nil
		} else {
			keys = keys[query.Offset:]
		}
	}
	if query.Limit >= 0 && query.Limit < len(keys) {
		keys = keys[:query.Limit]
	}

	responseRows := make([]tableDataRow, 0, len(keys))
	for _, key := range keys {
		aggregate, err := executeAggregateQuery(groups[key], schema, simpleSelectQuery{Aggregate: query.Aggregate})
		if err != nil {
			return queryExecutionResult{}, err
		}
		cell := tableCell{V: nil}
		if raw := groupValues[key]; !isJSONNull(raw) {
			cell.V = rawValueForFieldResponse(raw, fields[0])
		}
		if len(aggregate.Rows) == 0 || len(aggregate.Rows[0].F) == 0 {
			return queryExecutionResult{}, fmt.Errorf("aggregate result is empty")
		}
		responseRows = append(responseRows, tableDataRow{
			F: []tableCell{cell, aggregate.Rows[0].F[0]},
		})
	}
	return queryExecutionResult{Fields: fields, Rows: responseRows}, nil
}

func singleAggregateResult(name string, fieldType string, value string) queryExecutionResult {
	data, _ := json.Marshal(value)
	return queryExecutionResult{
		Fields: []tableFieldSchema{{
			Name: name,
			Type: fieldType,
			Mode: "NULLABLE",
		}},
		Rows: []tableDataRow{{
			F: []tableCell{{V: rawValueForResponse(data)}},
		}},
	}
}

func singleAggregateNullResult(name string, fieldType string) queryExecutionResult {
	return queryExecutionResult{
		Fields: []tableFieldSchema{{
			Name: name,
			Type: fieldType,
			Mode: "NULLABLE",
		}},
		Rows: []tableDataRow{{
			F: []tableCell{{V: nil}},
		}},
	}
}

func sumAggregate(rows []storedRow, fieldName string, integer bool) (string, int, error) {
	if integer {
		var sum int64
		count := 0
		for _, row := range rows {
			raw, ok := row.JSON[fieldName]
			if !ok || isJSONNull(raw) {
				continue
			}
			value, ok := rawInt(raw)
			if !ok {
				return "", 0, fmt.Errorf("SUM field contains a non-integer value")
			}
			sum += value
			count++
		}
		return strconv.FormatInt(sum, 10), count, nil
	}
	sum, count, err := floatAggregate(rows, fieldName)
	if err != nil {
		return "", 0, err
	}
	return strconv.FormatFloat(sum, 'f', -1, 64), count, nil
}

func floatAggregate(rows []storedRow, fieldName string) (float64, int, error) {
	var sum float64
	count := 0
	for _, row := range rows {
		raw, ok := row.JSON[fieldName]
		if !ok || isJSONNull(raw) {
			continue
		}
		value, ok := rawFloat(raw)
		if !ok {
			return 0, 0, fmt.Errorf("aggregate field contains a non-numeric value")
		}
		sum += value
		count++
	}
	return sum, count, nil
}

func minMaxAggregate(rows []storedRow, fieldName string, function string) (json.RawMessage, bool) {
	var selected json.RawMessage
	found := false
	for _, row := range rows {
		raw, ok := row.JSON[fieldName]
		if !ok || isJSONNull(raw) {
			continue
		}
		if !found {
			selected = raw
			found = true
			continue
		}
		cmp := compareRawValues(raw, selected)
		if function == "MIN" && cmp < 0 || function == "MAX" && cmp > 0 {
			selected = raw
		}
	}
	return selected, found
}

func isNumericField(field tableFieldSchema) bool {
	fieldType := strings.ToUpper(defaultString(field.Type, "STRING"))
	switch fieldType {
	case "INTEGER", "INT64", "FLOAT", "FLOAT64", "NUMERIC", "BIGNUMERIC":
		return true
	default:
		return false
	}
}

func isIntegerField(field tableFieldSchema) bool {
	fieldType := strings.ToUpper(defaultString(field.Type, "STRING"))
	return fieldType == "INTEGER" || fieldType == "INT64"
}

func aggregateNumericType(field tableFieldSchema) string {
	if isIntegerField(field) {
		return "INTEGER"
	}
	fieldType := strings.ToUpper(defaultString(field.Type, "FLOAT"))
	if fieldType == "FLOAT64" {
		return "FLOAT"
	}
	return fieldType
}

func rowMatchesQuery(row storedRow, query simpleSelectQuery) (bool, error) {
	if len(query.WhereConditionGroups) == 0 {
		if len(query.WhereConditions) > 0 {
			query.WhereConditionGroups = [][]whereCondition{query.WhereConditions}
		} else if query.WhereOperator != "" {
			query.WhereConditionGroups = [][]whereCondition{{
				{
					Field:    query.WhereField,
					Operator: query.WhereOperator,
					ValueRaw: query.WhereValueRaw,
				},
			}}
		}
	}
	if len(query.WhereConditionGroups) == 0 {
		return true, nil
	}
	for _, group := range query.WhereConditionGroups {
		matches, err := rowMatchesAllConditions(row, group)
		if err != nil {
			return false, err
		}
		if matches {
			return true, nil
		}
	}
	return false, nil
}

func rowMatchesAllConditions(row storedRow, conditions []whereCondition) (bool, error) {
	if len(conditions) == 0 {
		return true, nil
	}
	for _, condition := range conditions {
		raw, ok := row.JSON[condition.Field]
		if !ok || isJSONNull(raw) {
			return false, nil
		}
		cmp := compareRawValues(raw, condition.ValueRaw)
		var matches bool
		switch condition.Operator {
		case "=":
			matches = cmp == 0
		case "NOT =":
			matches = cmp != 0
		case "!=":
			matches = cmp != 0
		case "NOT !=":
			matches = cmp == 0
		case ">":
			matches = cmp > 0
		case "NOT >":
			matches = cmp <= 0
		case ">=":
			matches = cmp >= 0
		case "NOT >=":
			matches = cmp < 0
		case "<":
			matches = cmp < 0
		case "NOT <":
			matches = cmp >= 0
		case "<=":
			matches = cmp <= 0
		case "NOT <=":
			matches = cmp > 0
		default:
			return false, fmt.Errorf("unsupported WHERE operator")
		}
		if !matches {
			return false, nil
		}
	}
	return true, nil
}

func compareRawValues(left json.RawMessage, right json.RawMessage) int {
	leftNumber, leftNumberOK := rawFloat(left)
	rightNumber, rightNumberOK := rawFloat(right)
	if leftNumberOK && rightNumberOK {
		switch {
		case leftNumber < rightNumber:
			return -1
		case leftNumber > rightNumber:
			return 1
		default:
			return 0
		}
	}
	leftString := fmt.Sprint(rawValueForResponse(left))
	rightString := fmt.Sprint(rawValueForResponse(right))
	return strings.Compare(leftString, rightString)
}

func rawFloat(raw json.RawMessage) (float64, bool) {
	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		value, err := strconv.ParseFloat(asString, 64)
		return value, err == nil
	}
	var number json.Number
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	if err := decoder.Decode(&number); err != nil {
		return 0, false
	}
	value, err := strconv.ParseFloat(number.String(), 64)
	return value, err == nil
}

func rawInt(raw json.RawMessage) (int64, bool) {
	var asString string
	if err := json.Unmarshal(raw, &asString); err == nil {
		value, err := strconv.ParseInt(asString, 10, 64)
		return value, err == nil
	}
	var number json.Number
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	if err := decoder.Decode(&number); err != nil {
		return 0, false
	}
	value, err := strconv.ParseInt(number.String(), 10, 64)
	return value, err == nil
}

File: services/bigquery/src/sql_eval.rs
//! Expression/predicate evaluation and aggregates — port of
//! `internal/services/bigquery/sql_eval.rs`, plus the raw-JSON cell helpers
//! from `tabledata_handlers.rs` (`isJSONNull`, `formatRowValues`,
//! `rawValueForResponse`, `rawValueForFieldResponse`) that the engine shares
//! with the tabledata handlers (part 3).
//!
//! legacy decodes raw JSON with `UseNumber`, so cell values carry `json.Number`
//! and re-encode as the original literal. `serde_json::Value` (without
//! `arbitrary_precision`) normalizes the number text instead (`1.50` → `1.5`,
//! `1e2` → `100.0`); integers and canonically-formatted floats — everything
//! the legacy test suite exercises — round-trip identically.

use std::cmp::Ordering;
use std::collections::BTreeMap;

use serde_json::Value;

use crate::model::{RawJson, StoredRow, TableCell, TableDataRow, TableFieldSchema, TableSchema};
use crate::responses::default_string;
use crate::sql_parser::{
    aggregate_field, fields_for_query, grouped_aggregate_dry_run_fields, AggregateFunction,
    AggregateSelection, ComparisonOp, SimpleSelectQuery, WhereCondition,
};
use crate::wire_json;

/// legacy `queryExecutionResult`.
#[derive(Debug, Clone, Default)]
pub struct QueryExecutionResult {
    pub fields: Vec<TableFieldSchema>,
    pub rows: Vec<TableDataRow>,
}

/// legacy `executeParsedQuery`.
pub fn execute_parsed_query(
    schema: &TableSchema,
    rows: &[StoredRow],
    query: &SimpleSelectQuery,
) -> Result<QueryExecutionResult, String> {
    let selected_fields = fields_for_query(schema, &query.selected_fields)?;
    let mut filtered: Vec<&StoredRow> = rows
        .iter()
        .filter(|row| row_matches_query(row, query))
        .collect();
    if let Some(aggregate) = &query.aggregate {
        if !query.group_by.is_empty() {
            return execute_grouped_aggregate_query(&filtered, schema, query, aggregate);
        }
        return execute_aggregate_query(&filtered, schema, aggregate);
    }
    if !query.order_by.is_empty() {
        filtered.sort_by(|a, b| {
            let cmp =
                compare_raw_values(raw_field(a, &query.order_by), raw_field(b, &query.order_by));
            if query.order_desc {
                cmp.reverse()
            } else {
                cmp
            }
        });
    }
    if query.offset > 0 {
        if query.offset >= filtered.len() as i64 {
            filtered.clear();
        } else {
            filtered.drain(..query.offset as usize);
        }
    }
    if query.limit >= 0 && query.limit < filtered.len() as i64 {
        filtered.truncate(query.limit as usize);
    }
    let response_rows = filtered
        .iter()
        .map(|row| TableDataRow {
            f: format_row_values(&row.json, &selected_fields),
        })
        .collect();
    Ok(QueryExecutionResult {
        fields: selected_fields,
        rows: response_rows,
    })
}

/// Missing key ≙ legacy nil `json.RawMessage` (empty bytes): every decode fails,
/// numeric comparison falls back to the empty string.
fn raw_field<'a>(row: &'a StoredRow, name: &str) -> &'a str {
    row.json.get(name).map(|raw| raw.get()).unwrap_or("")
}

/// legacy `queryResultRowsToStoredRows`.
pub fn query_result_rows_to_stored_rows(result: &QueryExecutionResult) -> Vec<StoredRow> {
    result
        .rows
        .iter()
        .map(|row| {
            let mut values: BTreeMap<String, RawJson> = BTreeMap::new();
            for (i, field) in result.fields.iter().enumerate() {
                let value = row
                    .f
                    .get(i)
                    .map(|cell| cell.v.clone())
                    .unwrap_or(Value::Null);
                let raw =
                    String::from_utf8(wire_json::marshal(&value)).expect("marshaled json is utf-8");
                let raw = serde_json::value::RawValue::from_string(raw)
                    .expect("marshaled json is a valid raw value");
                values.insert(field.name.clone(), raw);
            }
            StoredRow {
                insert_id: String::new(),
                json: values,
                inserted_at: String::new(),
            }
        })
        .collect()
}

/// legacy `executeAggregateQuery`.
pub(crate) fn execute_aggregate_query(
    rows: &[&StoredRow],
    schema: &TableSchema,
    aggregate: &AggregateSelection,
) -> Result<QueryExecutionResult, String> {
    let field_name = if aggregate.alias.is_empty() {
        "f0_"
    } else {
        &aggregate.alias
    };

    let field = aggregate_field(schema, aggregate);

    if aggregate.field == "*" {
        if aggregate.function != AggregateFunction::Count {
            return Err(format!("{} requires a field", aggregate.function.name()));
        }
        return Ok(single_aggregate_result(
            field_name,
            "INTEGER",
            &rows.len().to_string(),
        ));
    }
    let Some(field) = field else {
        return Err(format!(
            "aggregate field {:?} does not exist",
            aggregate.field
        ));
    };

    match aggregate.function {
        AggregateFunction::Count => {
            let count = rows
                .iter()
                .filter(|row| {
                    row.json
                        .get(&aggregate.field)
                        .map(|raw| !is_json_null(raw.get()))
                        .unwrap_or(false)
                })
                .count();
            Ok(single_aggregate_result(
                field_name,
                "INTEGER",
                &count.to_string(),
            ))
        }
        AggregateFunction::Sum => {
            if !is_numeric_field(field) {
                return Err("SUM requires a numeric field".to_string());
            }
            let (sum, count) = sum_aggregate(rows, &aggregate.field, is_integer_field(field))?;
            if count == 0 {
                return Ok(single_aggregate_null_result(
                    field_name,
                    &aggregate_numeric_type(field),
                ));
            }
            Ok(single_aggregate_result(
                field_name,
                &aggregate_numeric_type(field),
                &sum,
            ))
        }
        AggregateFunction::Avg => {
            if !is_numeric_field(field) {
                return Err("AVG requires a numeric field".to_string());
            }
            let (sum, count) = float_aggregate(rows, &aggregate.field)?;
            if count == 0 {
                return Ok(single_aggregate_null_result(field_name, "FLOAT"));
            }
            Ok(single_aggregate_result(
                field_name,
                "FLOAT",
                &format_float_legacy(sum / count as f64),
            ))
        }
        AggregateFunction::Min | AggregateFunction::Max => {
            let field_type = default_string(field.field_type.clone(), "STRING");
            let Some(raw) = min_max_aggregate(rows, &aggregate.field, aggregate.function) else {
                return Ok(single_aggregate_null_result(field_name, &field_type));
            };
            Ok(QueryExecutionResult {
                fields: vec![TableFieldSchema {
                    name: field_name.to_string(),
                    field_type,
                    mode: "NULLABLE".to_string(),
                    ..Default::default()
                }],
                rows: vec![TableDataRow {
                    f: vec![TableCell {
                        v: raw_value_for_response(raw),
                    }],
                }],
            })
        }
    }
}

/// legacy `executeGroupedAggregateQuery`.
fn execute_grouped_aggregate_query(
    rows: &[&StoredRow],
    schema: &TableSchema,
    query: &SimpleSelectQuery,
    aggregate: &AggregateSelection,
) -> Result<QueryExecutionResult, String> {
    let fields = grouped_aggregate_dry_run_fields(schema, &query.group_by, aggregate)?;
    // Groups are keyed by the raw JSON text of the grouped value (missing →
    // "null"), like legacy `string(raw)` map key. legacy sorts the keys with a
    // non-stable sort over random map iteration order; keys are unique, so the
    // only ties `compareRawValues` can report are numerically-equal distinct
    // literals, where legacy order is unspecified — first-seen stable order here.
    let mut keys: Vec<String> = Vec::new();
    let mut groups: BTreeMap<String, Vec<&StoredRow>> = BTreeMap::new();
    for row in rows {
        let raw = row
            .json
            .get(&query.group_by)
            .map(|raw| raw.get())
            .unwrap_or("null");
        if !groups.contains_key(raw) {
            keys.push(raw.to_string());
        }
        groups.entry(raw.to_string()).or_default().push(row);
    }
    keys.sort_by(|a, b| {
        let cmp = compare_raw_values(a, b);
        if query.order_desc {
            cmp.reverse()
        } else {
            cmp
        }
    });
    if !query.order_by.is_empty() && query.order_by != query.group_by {
        return Err("ORDER BY supports grouped field only for GROUP BY queries".to_string());
    }
    if query.offset > 0 {
        if query.offset >= keys.len() as i64 {
            keys.clear();
        } else {
            keys.drain(..query.offset as usize);
        }
    }
    if query.limit >= 0 && query.limit < keys.len() as i64 {
        keys.truncate(query.limit as usize);
    }

    let mut response_rows = Vec::with_capacity(keys.len());
    for key in &keys {
        let aggregate_result = execute_aggregate_query(&groups[key], schema, aggregate)?;
        let cell = TableCell {
            v: if is_json_null(key) {
                Value::Null
            } else {
                raw_value_for_field_response(key, &fields[0])
            },
        };
        let Some(aggregate_cell) = aggregate_result
            .rows
            .first()
            .and_then(|row| row.f.first())
            .cloned()
        else {
            return Err("aggregate result is empty".to_string());
        };
        response_rows.push(TableDataRow {
            f: vec![cell, aggregate_cell],
        });
    }
    Ok(QueryExecutionResult {
        fields,
        rows: response_rows,
    })
}

/// legacy `singleAggregateResult` (the value is always rendered as a JSON string
/// cell; legacy marshals then re-decodes it, which is the identity).
fn single_aggregate_result(name: &str, field_type: &str, value: &str) -> QueryExecutionResult {
    QueryExecutionResult {
        fields: vec![TableFieldSchema {
            name: name.to_string(),
            field_type: field_type.to_string(),
            mode: "NULLABLE".to_string(),
            ..Default::default()
        }],
        rows: vec![TableDataRow {
            f: vec![TableCell {
                v: Value::String(value.to_string()),
            }],
        }],
    }
}

/// legacy `singleAggregateNullResult`.
fn single_aggregate_null_result(name: &str, field_type: &str) -> QueryExecutionResult {
    QueryExecutionResult {
        fields: vec![TableFieldSchema {
            name: name.to_string(),
            field_type: field_type.to_string(),
            mode: "NULLABLE".to_string(),
            ..Default::default()
        }],
        rows: vec![TableDataRow {
            f: vec![TableCell { v: Value::Null }],
        }],
    }
}

/// legacy `sumAggregate`. Integer sums wrap on overflow like legacy `int64 +=`.
fn sum_aggregate(
    rows: &[&StoredRow],
    field_name: &str,
    integer: bool,
) -> Result<(String, usize), String> {
    if integer {
        let mut sum: i64 = 0;
        let mut count = 0;
        for row in rows {
            let Some(raw) = row.json.get(field_name) else {
                continue;
            };
            if is_json_null(raw.get()) {
                continue;
            }
            let Some(value) = raw_int(raw.get()) else {
                return Err("SUM field contains a non-integer value".to_string());
            };
            sum = sum.wrapping_add(value);
            count += 1;
        }
        return Ok((sum.to_string(), count));
    }
    let (sum, count) = float_aggregate(rows, field_name)?;
    Ok((format_float_legacy(sum), count))
}

/// legacy `floatAggregate`.
fn float_aggregate(rows: &[&StoredRow], field_name: &str) -> Result<(f64, usize), String> {
    let mut sum = 0.0;
    let mut count = 0;
    for row in rows {
        let Some(raw) = row.json.get(field_name) else {
            continue;
        };
        if is_json_null(raw.get()) {
            continue;
        }
        let Some(value) = raw_float(raw.get()) else {
            return Err("aggregate field contains a non-numeric value".to_string());
        };
        sum += value;
        count += 1;
    }
    Ok((sum, count))
}

/// legacy `minMaxAggregate`.
fn min_max_aggregate<'a>(
    rows: &[&'a StoredRow],
    field_name: &str,
    function: AggregateFunction,
) -> Option<&'a str> {
    let mut selected: Option<&str> = None;
    for row in rows {
        let Some(raw) = row.json.get(field_name) else {
            continue;
        };
        let raw = raw.get();
        if is_json_null(raw) {
            continue;
        }
        let Some(current) = selected else {
            selected = Some(raw);
            continue;
        };
        let cmp = compare_raw_values(raw, current);
        if function == AggregateFunction::Min && cmp == Ordering::Less
            || function == AggregateFunction::Max && cmp == Ordering::Greater
        {
            selected = Some(raw);
        }
    }
    selected
}

/// legacy `isNumericField`.
pub(crate) fn is_numeric_field(field: &TableFieldSchema) -> bool {
    matches!(
        default_string(field.field_type.clone(), "STRING")
            .to_uppercase()
            .as_str(),
        "INTEGER" | "INT64" | "FLOAT" | "FLOAT64" | "NUMERIC" | "BIGNUMERIC"
    )
}

/// legacy `isIntegerField`.
pub(crate) fn is_integer_field(field: &TableFieldSchema) -> bool {
    matches!(
        default_string(field.field_type.clone(), "STRING")
            .to_uppercase()
            .as_str(),
        "INTEGER" | "INT64"
    )
}

/// legacy `aggregateNumericType`.
pub(crate) fn aggregate_numeric_type(field: &TableFieldSchema) -> String {
    if is_integer_field(field) {
        return "INTEGER".to_string();
    }
    let field_type = default_string(field.field_type.clone(), "FLOAT").to_uppercase();
    if field_type == "FLOAT64" {
        "FLOAT".to_string()
    } else {
        field_type
    }
}

/// legacy `rowMatchesQuery`. The legacy fallbacks that rebuild the groups from the
/// flattened `WhereConditions` / legacy single-condition fields are
/// unreachable from the parser (it always fills the groups), so only the
/// groups are consulted.
fn row_matches_query(row: &StoredRow, query: &SimpleSelectQuery) -> bool {
    if query.where_condition_groups.is_empty() {
        return true;
    }
    query
        .where_condition_groups
        .iter()
        .any(|group| row_matches_all_conditions(row, group))
}

/// legacy `rowMatchesAllConditions`.
fn row_matches_all_conditions(row: &StoredRow, conditions: &[WhereCondition]) -> bool {
    conditions.iter().all(|condition| {
        let Some(raw) = row.json.get(&condition.field) else {
            return false;
        };
        let raw = raw.get();
        if is_json_null(raw) {
            return false;
        }
        let cmp = compare_raw_values(raw, &condition.value_raw);
        let matches = match condition.op {
            ComparisonOp::Eq => cmp == Ordering::Equal,
            ComparisonOp::NotEq => cmp != Ordering::Equal,
            ComparisonOp::Gt => cmp == Ordering::Greater,
            ComparisonOp::Ge => cmp != Ordering::Less,
            ComparisonOp::Lt => cmp == Ordering::Less,
            ComparisonOp::Le => cmp != Ordering::Greater,
        };
        // legacy spells every `NOT <op>` case out; each is the exact negation.
        matches != condition.negated
    })
}

/// legacy `compareRawValues`: numeric comparison when both sides parse as floats
/// (NaN compares equal, like legacy `<`/`>` chain), otherwise byte-wise string
/// comparison of the `fmt.Sprint`-rendered decoded values.
pub(crate) fn compare_raw_values(left: &str, right: &str) -> Ordering {
    if let (Some(left_number), Some(right_number)) = (raw_float(left), raw_float(right)) {
        return left_number
            .partial_cmp(&right_number)
            .unwrap_or(Ordering::Equal);
    }
    legacy_sprint(&raw_value_for_response(left)).cmp(&legacy_sprint(&raw_value_for_response(right)))
}

/// legacy `rawFloat`: a JSON string parsed with `ParseFloat`, or a JSON number.
pub(crate) fn raw_float(raw: &str) -> Option<f64> {
    if let Ok(as_string) = serde_json::from_str::<String>(raw) {
        return as_string.parse::<f64>().ok();
    }
    match first_json_value(raw) {
        Some(Value::Number(number)) => number.as_f64(),
        _ => None,
    }
}

/// legacy `rawInt`.
pub(crate) fn raw_int(raw: &str) -> Option<i64> {
    if let Ok(as_string) = serde_json::from_str::<String>(raw) {
        return as_string.parse::<i64>().ok();
    }
    match first_json_value(raw) {
        Some(Value::Number(number)) => number.as_i64(),
        _ => None,
    }
}

/// legacy `isJSONNull` (tabledata_handlers.rs).
pub(crate) fn is_json_null(raw: &str) -> bool {
    raw.trim().eq_ignore_ascii_case("null")
}

/// legacy `formatRowValues` (tabledata_handlers.rs).
pub(crate) fn format_row_values(
    row: &BTreeMap<String, RawJson>,
    fields: &[TableFieldSchema],
) -> Vec<TableCell> {
    fields
        .iter()
        .map(|field| match row.get(&field.name) {
            Some(raw) if !is_json_null(raw.get()) => TableCell {
                v: raw_value_for_field_response(raw.get(), field),
            },
            _ => TableCell { v: Value::Null },
        })
        .collect()
}

/// legacy `rawValueForFieldResponse` (tabledata_handlers.rs): INTEGER/BOOLEAN
/// (and NUMERIC) cells render as strings, everything else decodes as-is.
pub(crate) fn raw_value_for_field_response(raw: &str, field: &TableFieldSchema) -> Value {
    if field.mode.eq_ignore_ascii_case("REPEATED") {
        return raw_value_for_response(raw);
    }
    match default_string(field.field_type.clone(), "STRING")
        .to_uppercase()
        .as_str()
    {
        "INTEGER" | "INT64" => {
            if let Some(value) = raw_int(raw) {
                return Value::String(value.to_string());
            }
        }
        "BOOLEAN" | "BOOL" => {
            if let Ok(value) = serde_json::from_str::<bool>(raw) {
                return Value::String(value.to_string());
            }
        }
        "NUMERIC" | "BIGNUMERIC" => {
            if let Ok(value) = serde_json::from_str::<String>(raw) {
                return Value::String(value);
            }
            if let Some(Value::Number(number)) = first_json_value(raw) {
                return Value::String(number.to_string());
            }
        }
        _ => {}
    }
    raw_value_for_response(raw)
}

/// legacy `rawValueForResponse` (tabledata_handlers.rs): decode the first JSON
/// value (trailing bytes ignored, like a single `Decoder.Decode` call);
/// undecodable bytes come back as their literal text.
pub(crate) fn raw_value_for_response(raw: &str) -> Value {
    match first_json_value(raw) {
        Some(value) => value,
        None => Value::String(raw.to_string()),
    }
}

fn first_json_value(raw: &str) -> Option<Value> {
    match serde_json::Deserializer::from_str(raw)
        .into_iter::<Value>()
        .next()
    {
        Some(Ok(value)) => Some(value),
        _ => None,
    }
}

/// legacy `fmt.Sprint` over a `UseNumber`-decoded JSON value: `nil` → `<nil>`,
/// slices space-joined in brackets, maps as `map[k:v ...]` with sorted keys
/// (serde's `Value` object map is a `BTreeMap`, matching legacy sorted `Sprint`).
pub(crate) fn legacy_sprint(value: &Value) -> String {
    match value {
        Value::Null => "<nil>".to_string(),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        Value::String(s) => s.clone(),
        Value::Array(items) => {
            let parts: Vec<String> = items.iter().map(legacy_sprint).collect();
            format!("[{}]", parts.join(" "))
        }
        Value::Object(map) => {
            let parts: Vec<String> = map
                .iter()
                .map(|(key, value)| format!("{key}:{}", legacy_sprint(value)))
                .collect();
            format!("map[{}]", parts.join(" "))
        }
    }
}

/// legacy `strconv.FormatFloat(value, 'f', -1, 64)`: shortest round-trip decimal,
/// never scientific notation — Rust's `Display` for `f64` matches, except for
/// the non-finite spellings.
fn format_float_legacy(value: f64) -> String {
    if value.is_nan() {
        return "NaN".to_string();
    }
    if value.is_infinite() {
        return if value > 0.0 { "+Inf" } else { "-Inf" }.to_string();
    }
    format!("{value}")
}
